Python/ast_opt.c

```diff
@@ -0,0 +1,649 @@
+/* AST Optimizer */
+#include "Python.h"
+#include "Python-ast.h"
+
+
+/* TODO: is_const and get_const_value are copied from Python/compile.c.
+   It should be deduped in the future.  Maybe, we can include this file
+   from compile.c?
+*/
+static int
+is_const(expr_ty e)
+{
+    switch (e->kind) {
+    case Constant_kind:
+    case Num_kind:
+    case Str_kind:
+    case Bytes_kind:
+    case Ellipsis_kind:
+    case NameConstant_kind:
+        return 1;
+    default:
+        return 0;
+    }
+}
+
+static PyObject *
+get_const_value(expr_ty e)
+{
+    switch (e->kind) {
+    case Constant_kind:
+        return e->v.Constant.value;
+    case Num_kind:
+        return e->v.Num.n;
+    case Str_kind:
+        return e->v.Str.s;
+    case Bytes_kind:
+        return e->v.Bytes.s;
+    case Ellipsis_kind:
+        return Py_Ellipsis;
+    case NameConstant_kind:
+        return e->v.NameConstant.value;
+    default:
+        Py_UNREACHABLE();
+    }
+}
+
+static int
+make_const(expr_ty node, PyObject *val, PyArena *arena)
+{
+    if (val == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_KeyboardInterrupt)) {
+            return 0;
+        }
+        PyErr_Clear();
+        return 1;
+    }
+    if (PyArena_AddPyObject(arena, val) < 0) {
+        Py_DECREF(val);
+        return 0;
+    }
+    node->kind = Constant_kind;
+    node->v.Constant.value = val;
+    return 1;
+}
+
+#define COPY_NODE(TO, FROM) (memcpy((TO), (FROM), sizeof(struct _expr)))
+
+static PyObject*
+unary_not(PyObject *v)
+{
+    int r = PyObject_IsTrue(v);
+    if (r < 0)
+        return NULL;
+    return PyBool_FromLong(!r);
+}
+
+static int
+fold_unaryop(expr_ty node, PyArena *arena)
+{
+    expr_ty arg = node->v.UnaryOp.operand;
+
+    if (!is_const(arg)) {
+        /* Fold not into comparison */
```

Comment:
Looks like this makes [this](https://github.com/methane/cpython/blob/30edda97b4acd07da7a16c3de7726cd5a23148a9/Python/peephole.c#L348) peephole pass unnecessary too?

Reply:
Ah, you're right.

Reply:
Created #4863.  Thanks.